upgrader.go

```diff
@@ -106,12 +107,20 @@ func (u *Upgrader) Upgrade(ctx context.Context, t transport.Transport, maconn ma
 	// call the connection gater, if one is registered.
 	if u.ConnGater != nil && !u.ConnGater.InterceptSecured(dir, sconn.RemotePeer(), maconn) {
 		if err := maconn.Close(); err != nil {
-			log.Errorf("failed to close connection with peer %s and addr %s; err: %s",
-				p.Pretty(), maconn.RemoteMultiaddr(), err)
+			log.Errorw("failed to close connection", "peer", p, "addr", maconn.RemoteMultiaddr(), "error", err)
 		}
 		return nil, fmt.Errorf("gater rejected connection with peer %s and addr %s with direction %d",
 			sconn.RemotePeer().Pretty(), maconn.RemoteMultiaddr(), dir)
 	}
+	if p == "" && connScope != nil {
+		if err := connScope.SetPeer(sconn.RemotePeer()); err != nil {
```

Comment:
we need a Done in the error path, as we own it here.

Reply:
so where is this getting Done? The caller is responsible?

Reply:
In the swarm: https://github.com/libp2p/go-libp2p-swarm/pull/308/files#diff-decb5d093ca975d319efcad30e802c35efa8f38ede82777bbd206239ca01b9bdR97